File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
      /* Approach One
        //using HashMap 
        //tc:O(N) and sc:O(N)
       Map<Integer, Integer> map = new HashMap<>();

	for (int i = 0; i < nums.length; i++) {
		if (map.containsKey(nums[i]) &&(i - map.get(nums[i]) <= k))
		return true;
		map.put(nums[i], i);
	}

	return false;
        */
        //Approach Two
        // sliding window and HashSet
        //tc:O(N) and sc:O(N)
        Set<Integer> slidingWindowHash = new HashSet();
        for (int i = 0; i < nums.length; i++) {
            if (slidingWindowHash.contains(nums[i])) {
                return true;
            }
            slidingWindowHash.add(nums[i]);
            
            if (i >= k) {
                slidingWindowHash.remove(nums[i - k]);
            }
        }
        return false;
    }
}

File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        # for i in range(len(nums)-1):
        #     for j in range(i+1,len(nums)):
        #         if nums[i]==nums[j] and abs(i-j)<=k:
        #             return True
        # return False
        left=0
        set_container=set()
        for right in range(len(nums)):
            if nums[right] in set_container:
                return True
            set_container.add(nums[right])
            if right-left==k:
                set_container.remove(nums[left])
                left+=1
        return False